conda/resolve.py

```diff
@@ -357,167 +325,181 @@ def verify_specs(self, specs):
+                  ['numpy 1.5*', 'scipy 0.12.0b1']
+            - Each spec depends on *the same package* but in a different way; e.g.,
+                  ['A', 'B'] where A depends on numpy 1.5, and B on numpy 1.6.
+            Technically, all three of these cases can be boiled down to the last
+            one if we treat the spec itself as one of the "dependencies". There
+            might be more complex reasons for a conflict, but this code only
+            considers the ones above.
+
+            The purpose of this code, then, is to identify packages (like numpy
+            above) that all of the specs depend on *but in different ways*. We
+            then identify the dependency chains that lead to those packages.
+        """
+        sdeps = {}
+        # For each spec, assemble a dictionary of dependencies, with package
+        # name as key, and all of the matching packages as values.
+        for ms in specs:
+            rec = sdeps.setdefault(ms, {})
+            slist = [ms]
+            while slist:
+                ms2 = slist.pop()
+                deps = rec.setdefault(ms2.name, set())
+                for fkey in self.find_matches(ms2):
+                    if fkey not in deps:
+                        deps.add(fkey)
+                        slist.extend(ms3 for ms3 in self.ms_depends(fkey) if ms3.name != ms.name)
+        # Find the list of dependencies they have in common. And for each of
+        # *those*, find the individual packages that they all share. Those need
+        # to be removed as conflict candidates.
+        commkeys = set.intersection(*(set(s.keys()) for s in sdeps.values()))
+        commkeys = {k: set.intersection(*(v[k] for v in sdeps.values())) for k in commkeys}
+        # and find the dependency chains that lead to them.
+        bad_deps = []
+        for ms, sdep in iteritems(sdeps):
+            filter = {}
+            for mn, v in sdep.items():
+                if mn != ms.name and mn in commkeys:
+                    # Mark this package's "unique" dependencies as invali
+                    for fkey in v - commkeys[mn]:
+                        filter[fkey] = False
+            # Find the dependencies that lead to those invalid choices
+            ndeps = set(self.invalid_chains(ms, filter))
+            # This may produce some additional invalid chains that we
+            # don't care about. Select only those that terminate in our
+            # predetermined set of "common" keys.
+            ndeps = [nd for nd in ndeps if nd[-1].name in commkeys]
+            if ndeps:
+                bad_deps.extend(ndeps)
+            else:
+                # This means the package *itself* was the common conflict.
+                bad_deps.append((ms,))
+        raise UnsatisfiableError(bad_deps)
 
     def get_dists(self, specs):
-        log.debug('Retrieving packages for: %s' % specs)
+        log.debug('Retrieving packages for: %s' % (specs,))
```

Comment:
When calling log, best practice is to avoid string interpolation at call time.  `log` supports string interpolation with extra `*args` to delay building the final string if it's not needed (i.e. most of the time in the `log.debug()` case).  Ideally this line would just be

    log.debug('Retrieving packages for: %s', specs)